### ISSUE TITLE
False negative: Nullness Checker allows leaking implicit `@UnderInitialization this` to inner class c'tor

### ISSUE THREAD

**Issue by PawelLipski** [typetools/checker-framework#3407]
Nullness Checker apparently does not verify the implicit reference to enclosing class instance.

Let's consider the following class:

```java
class SampleClass {
  final InnerClass inner;
  final String foo;

  SampleClass() {
    this.inner = new InnerClass();
    this.inner.bar(); /// WHOOPS... NPE
    this.foo = "Hello world";
  }

  class InnerClass {
    String bar() {
      return foo.substring(1);
    }
  }
}
```

This compiles with no error, despite clearly leading to an NPE in the runtime.

A nearly equivalent snippet that uses a static nested class instead:

```java
class SampleClass {
  final InnerClass inner;
  final String foo;

  SampleClass() {
    this.inner = new InnerClass(this); // Ok, NullnessChecker raises an error
    this.inner.bar();
    this.foo = "Hello world";
  }

  static class InnerClass {
    SampleClass outer;
    InnerClass(SampleClass outer) {
      this.outer = outer;
    }
    String bar() {
      return outer.foo.substring(1);
    }
  }
}
```

fails as expected on `nullness:assignment.type.incompatible` and `argument.type.incompatible`.

**Comment by mernst** [comment]
Possible duplicate of #409

**Comment by PawelLipski** [comment]
Hi @mernst, any prospect of this (and/or #409) getting fixed?... as for now, out of paranoia I've just added a [rule in an ArchUnit spec](https://github.com/VirtusLab/git-machete-intellij-plugin/blob/develop/src/test/java/com/virtuslab/archunit/ClassStructureTestSuite.java#L22) that bans all inner classes across the project, given I can't 100% reliably ensure that no NPE happens when they're initialized :sweat_smile:

**Comment by mernst** [comment]
@PawelLipski Sorry for the delay.  We are currently focused on fixing https://github.com/typetools/checker-framework/issues/979 (https://github.com/typetools/checker-framework/pull/3922 is a step toward that).  However, we will try to fit this bug fix in too.

**Comment by mkondratek** [comment]
+1 to this one
